source/loader/layers/validation/ur_valddi.cpp

```diff
@@ -9,7 +9,51 @@
  */
 #include "ur_validation_layer.hpp"
 
+#include <atomic>
+#include <functional>
+#include <map>
+#include <mutex>
+#include <shared_mutex>
+#include <utility>
+
 namespace validation_layer {
+struct {
+    std::map<void *, std::atomic<int>> counts;
+    std::shared_timed_mutex mutex;
+
+    std::pair<void *, int> incrementRefCount(void *ptr) {
+        std::shared_lock<std::shared_timed_mutex> lock(mutex);
+        auto it = counts.find(ptr);
+        if (it != counts.end()) {
+            lock.unlock();
+            std::unique_lock<std::shared_timed_mutex> ulock(mutex);
```

Comment:
There's a potential race here. After you call unlock, some other thread might call decrementRefCount() and remove the element pointed to by `it` before you take the `unique_lock`.

The simplest solution would be to just always use `std::mutex` and `std::unique_lock`. Or you can use something like https://en.wikipedia.org/wiki/Double-checked_locking if you want to optimize for existing elements. However, I'm not entirely sure if find() on std::map is actually thread-safe... (standard does not specify it I think and there might be some internal state updates even on find).

Reply:
It would be good if you added some multithreaded tests.

Reply:
Done.
I've omitted a case where I call urContextCreate by multiple threads because sometimes urContextCreate provides the same handle in different calls and it interferes with my .out.match tests. I'm not sure if urContextCreate is threadsafe.